From 1d7b9e97ca2d2c59ae9ff4c68dd87ce8ed4486a7 Mon Sep 17 00:00:00 2001
From: Chen Meng J <meng.j.chen@intel.com>
Date: Fri, 7 Jan 2022 12:31:02 +0800
Subject: [PATCH] media: imx390: lens shading correction for DSL369

Change Description:
Message for Open Source:
lens shading correction for DSL369.
then it can switch lsc pattern between lens 369 and 958.

Message for Internal:
lens shading correction for DSL369.

Tracked-On: #JIITL8-439
Signed-off-by: Chen Meng J <meng.j.chen@intel.com>

Change-Id: I0a77e020e31baf9409645bfd1ef3a9296ecb59ff
---
 drivers/media/i2c/imx390.c | 520 ++++++++++++++++++++++++++++++++++++-
 1 file changed, 512 insertions(+), 8 deletions(-)

diff --git a/drivers/media/i2c/imx390.c b/drivers/media/i2c/imx390.c
index c140be78a..ac8a730f7 100644
--- a/drivers/media/i2c/imx390.c
+++ b/drivers/media/i2c/imx390.c
@@ -291,13 +291,15 @@ static int imx390_group_hold_enable(struct imx390 *imx390, s32 val);
 enum {
 	LSC_PATTERN_DISABLED = 0,
 	LSC_PATTERN_UNITY,
-	LSC_PATTERN_TABLE,
+	LSC_PATTERN_TABLE_958,
+	LSC_PATTERN_TABLE_369,
 };
 
 static const char * const lsc_qmenu[] = {
 	"Disabled",
 	"all 0x80",
-	"table",
+	"lsc table for 958",
+	"lsc table for 369",
 };
 
 static const struct imx390_reg imx390_lsc_pattern_unity[] = {
@@ -524,7 +526,7 @@ static const struct imx390_reg_list lsc_unity_list = {
 	.regs = imx390_lsc_pattern_unity,
 };
 
-static const struct imx390_reg imx390_lsc_pattern_vendor_def[] = {
+static const struct imx390_reg imx390_lsc_pattern_958[] = {
 	{0x01D0, 0x01}, /* SHD_ON                    */
 	{0x3AF6, 0x00}, /* SHD_DIFF_ACCURACY         */
 	{0x4000, 0x64}, /* SHD_KNOT_1_R0             */
@@ -1018,9 +1020,508 @@ static const struct imx390_reg imx390_lsc_pattern_vendor_def[] = {
 	{0x384F, 0xFF}, /* OBB_CLAMP_OFFSET_B_SP2    */
 };
 
-static const struct imx390_reg_list lsc_vendor_def_list = {
-	.num_of_regs = ARRAY_SIZE(imx390_lsc_pattern_vendor_def),
-	.regs = imx390_lsc_pattern_vendor_def,
+static const struct imx390_reg_list lsc_958_list = {
+	.num_of_regs = ARRAY_SIZE(imx390_lsc_pattern_958),
+	.regs = imx390_lsc_pattern_958,
+};
+
+static const struct imx390_reg imx390_lsc_pattern_369[] = {
+	{0x01D0, 0x01},
+	{0x3AF6, 0x00},
+	{0x4000, 0xC6},
+	{0x4001, 0xA7},
+	{0x4002, 0xA8},
+	{0x4003, 0xA3},
+	{0x4004, 0xAB},
+	{0x4005, 0x98},
+	{0x4006, 0x98},
+	{0x4007, 0x95},
+	{0x4008, 0x9A},
+	{0x4009, 0x8E},
+	{0x400A, 0x8E},
+	{0x400B, 0x8D},
+	{0x400C, 0x8D},
+	{0x400D, 0x87},
+	{0x400E, 0x87},
+	{0x400F, 0x87},
+	{0x4010, 0x88},
+	{0x4011, 0x85},
+	{0x4012, 0x85},
+	{0x4013, 0x84},
+	{0x4014, 0x87},
+	{0x4015, 0x84},
+	{0x4016, 0x84},
+	{0x4017, 0x84},
+	{0x4018, 0x8A},
+	{0x4019, 0x86},
+	{0x401A, 0x86},
+	{0x401B, 0x85},
+	{0x401C, 0x90},
+	{0x401D, 0x8A},
+	{0x401E, 0x89},
+	{0x401F, 0x88},
+	{0x4020, 0x9B},
+	{0x4021, 0x91},
+	{0x4022, 0x90},
+	{0x4023, 0x8E},
+	{0x4024, 0xBB},
+	{0x4025, 0xA1},
+	{0x4026, 0xA1},
+	{0x4027, 0x9E},
+	{0x4028, 0xA5},
+	{0x4029, 0x95},
+	{0x402A, 0x95},
+	{0x402B, 0x93},
+	{0x402C, 0x90},
+	{0x402D, 0x88},
+	{0x402E, 0x88},
+	{0x402F, 0x87},
+	{0x4030, 0x87},
+	{0x4031, 0x84},
+	{0x4032, 0x84},
+	{0x4033, 0x83},
+	{0x4034, 0x82},
+	{0x4035, 0x81},
+	{0x4036, 0x81},
+	{0x4037, 0x81},
+	{0x4038, 0x82},
+	{0x4039, 0x81},
+	{0x403A, 0x81},
+	{0x403B, 0x81},
+	{0x403C, 0x85},
+	{0x403D, 0x82},
+	{0x403E, 0x82},
+	{0x403F, 0x82},
+	{0x4040, 0x8F},
+	{0x4041, 0x89},
+	{0x4042, 0x89},
+	{0x4043, 0x88},
+	{0x4044, 0x96},
+	{0x4045, 0x8E},
+	{0x4046, 0x8E},
+	{0x4047, 0x8C},
+	{0x4048, 0xB8},
+	{0x4049, 0x9F},
+	{0x404A, 0xA0},
+	{0x404B, 0x9C},
+	{0x404C, 0x9F},
+	{0x404D, 0x91},
+	{0x404E, 0x91},
+	{0x404F, 0x8F},
+	{0x4050, 0x8D},
+	{0x4051, 0x86},
+	{0x4052, 0x86},
+	{0x4053, 0x86},
+	{0x4054, 0x84},
+	{0x4055, 0x82},
+	{0x4056, 0x82},
+	{0x4057, 0x81},
+	{0x4058, 0x7F},
+	{0x4059, 0x7F},
+	{0x405A, 0x7F},
+	{0x405B, 0x7F},
+	{0x405C, 0x7F},
+	{0x405D, 0x7F},
+	{0x405E, 0x7F},
+	{0x405F, 0x80},
+	{0x4060, 0x83},
+	{0x4061, 0x82},
+	{0x4062, 0x82},
+	{0x4063, 0x82},
+	{0x4064, 0x8B},
+	{0x4065, 0x87},
+	{0x4066, 0x86},
+	{0x4067, 0x86},
+	{0x4068, 0x95},
+	{0x4069, 0x8D},
+	{0x406A, 0x8D},
+	{0x406B, 0x8C},
+	{0x406C, 0xBC},
+	{0x406D, 0xA2},
+	{0x406E, 0xA2},
+	{0x406F, 0x9D},
+	{0x4070, 0xA2},
+	{0x4071, 0x93},
+	{0x4072, 0x93},
+	{0x4073, 0x90},
+	{0x4074, 0x8E},
+	{0x4075, 0x87},
+	{0x4076, 0x87},
+	{0x4077, 0x86},
+	{0x4078, 0x85},
+	{0x4079, 0x82},
+	{0x407A, 0x82},
+	{0x407B, 0x82},
+	{0x407C, 0x81},
+	{0x407D, 0x80},
+	{0x407E, 0x80},
+	{0x407F, 0x80},
+	{0x4080, 0x81},
+	{0x4081, 0x81},
+	{0x4082, 0x81},
+	{0x4083, 0x81},
+	{0x4084, 0x85},
+	{0x4085, 0x83},
+	{0x4086, 0x83},
+	{0x4087, 0x83},
+	{0x4088, 0x8D},
+	{0x4089, 0x87},
+	{0x408A, 0x87},
+	{0x408B, 0x87},
+	{0x408C, 0x98},
+	{0x408D, 0x8E},
+	{0x408E, 0x8E},
+	{0x408F, 0x8D},
+	{0x4090, 0xC5},
+	{0x4091, 0xA7},
+	{0x4092, 0xA7},
+	{0x4093, 0xA2},
+	{0x4094, 0xAE},
+	{0x4095, 0x9B},
+	{0x4096, 0x9B},
+	{0x4097, 0x97},
+	{0x4098, 0x95},
+	{0x4099, 0x8B},
+	{0x409A, 0x8B},
+	{0x409B, 0x89},
+	{0x409C, 0x8B},
+	{0x409D, 0x85},
+	{0x409E, 0x85},
+	{0x409F, 0x85},
+	{0x40A0, 0x86},
+	{0x40A1, 0x83},
+	{0x40A2, 0x83},
+	{0x40A3, 0x83},
+	{0x40A4, 0x86},
+	{0x40A5, 0x83},
+	{0x40A6, 0x83},
+	{0x40A7, 0x83},
+	{0x40A8, 0x89},
+	{0x40A9, 0x84},
+	{0x40AA, 0x84},
+	{0x40AB, 0x84},
+	{0x40AC, 0x96},
+	{0x40AD, 0x8D},
+	{0x40AE, 0x8D},
+	{0x40AF, 0x8B},
+	{0x40B0, 0x9E},
+	{0x40B1, 0x92},
+	{0x40B2, 0x91},
+	{0x40B3, 0x90},
+	{0x40B4, 0xD6},
+	{0x40B5, 0xB1},
+	{0x40B6, 0xB1},
+	{0x40B7, 0xAA},
+	{0x40B8, 0xB8},
+	{0x40B9, 0x9F},
+	{0x40BA, 0x9F},
+	{0x40BB, 0x9A},
+	{0x40BC, 0xA4},
+	{0x40BD, 0x94},
+	{0x40BE, 0x93},
+	{0x40BF, 0x91},
+	{0x40C0, 0x96},
+	{0x40C1, 0x8C},
+	{0x40C2, 0x8C},
+	{0x40C3, 0x8A},
+	{0x40C4, 0x91},
+	{0x40C5, 0x89},
+	{0x40C6, 0x89},
+	{0x40C7, 0x88},
+	{0x40C8, 0x90},
+	{0x40C9, 0x89},
+	{0x40CA, 0x89},
+	{0x40CB, 0x88},
+	{0x40CC, 0x93},
+	{0x40CD, 0x8B},
+	{0x40CE, 0x8B},
+	{0x40CF, 0x8A},
+	{0x40D0, 0x9B},
+	{0x40D1, 0x8F},
+	{0x40D2, 0x8F},
+	{0x40D3, 0x8E},
+	{0x40D4, 0xA8},
+	{0x40D5, 0x97},
+	{0x40D6, 0x97},
+	{0x40D7, 0x95},
+	{0x4300, 0x9D},
+	{0x4301, 0x94},
+	{0x4302, 0x94},
+	{0x4303, 0x91},
+	{0x4304, 0x92},
+	{0x4305, 0x8C},
+	{0x4306, 0x8C},
+	{0x4307, 0x8B},
+	{0x4308, 0x8A},
+	{0x4309, 0x86},
+	{0x430A, 0x86},
+	{0x430B, 0x85},
+	{0x430C, 0x86},
+	{0x430D, 0x83},
+	{0x430E, 0x83},
+	{0x430F, 0x82},
+	{0x4310, 0x85},
+	{0x4311, 0x84},
+	{0x4312, 0x84},
+	{0x4313, 0x84},
+	{0x4314, 0x86},
+	{0x4315, 0x85},
+	{0x4316, 0x85},
+	{0x4317, 0x85},
+	{0x4318, 0x8B},
+	{0x4319, 0x89},
+	{0x431A, 0x89},
+	{0x431B, 0x89},
+	{0x431C, 0x91},
+	{0x431D, 0x8E},
+	{0x431E, 0x8E},
+	{0x431F, 0x8F},
+	{0x4320, 0x98},
+	{0x4321, 0x95},
+	{0x4322, 0x94},
+	{0x4323, 0x96},
+	{0x4324, 0x98},
+	{0x4325, 0x91},
+	{0x4326, 0x91},
+	{0x4327, 0x8F},
+	{0x4328, 0x8D},
+	{0x4329, 0x89},
+	{0x432A, 0x89},
+	{0x432B, 0x87},
+	{0x432C, 0x86},
+	{0x432D, 0x84},
+	{0x432E, 0x83},
+	{0x432F, 0x83},
+	{0x4330, 0x83},
+	{0x4331, 0x81},
+	{0x4332, 0x81},
+	{0x4333, 0x81},
+	{0x4334, 0x82},
+	{0x4335, 0x81},
+	{0x4336, 0x81},
+	{0x4337, 0x81},
+	{0x4338, 0x83},
+	{0x4339, 0x83},
+	{0x433A, 0x83},
+	{0x433B, 0x83},
+	{0x433C, 0x86},
+	{0x433D, 0x85},
+	{0x433E, 0x85},
+	{0x433F, 0x85},
+	{0x4340, 0x8C},
+	{0x4341, 0x8A},
+	{0x4342, 0x8A},
+	{0x4343, 0x8A},
+	{0x4344, 0x93},
+	{0x4345, 0x91},
+	{0x4346, 0x91},
+	{0x4347, 0x91},
+	{0x4348, 0x95},
+	{0x4349, 0x8E},
+	{0x434A, 0x8E},
+	{0x434B, 0x8D},
+	{0x434C, 0x8A},
+	{0x434D, 0x87},
+	{0x434E, 0x87},
+	{0x434F, 0x86},
+	{0x4350, 0x83},
+	{0x4351, 0x83},
+	{0x4352, 0x82},
+	{0x4353, 0x82},
+	{0x4354, 0x81},
+	{0x4355, 0x81},
+	{0x4356, 0x80},
+	{0x4357, 0x81},
+	{0x4358, 0x80},
+	{0x4359, 0x80},
+	{0x435A, 0x80},
+	{0x435B, 0x80},
+	{0x435C, 0x81},
+	{0x435D, 0x81},
+	{0x435E, 0x81},
+	{0x435F, 0x81},
+	{0x4360, 0x84},
+	{0x4361, 0x84},
+	{0x4362, 0x84},
+	{0x4363, 0x83},
+	{0x4364, 0x89},
+	{0x4365, 0x89},
+	{0x4366, 0x88},
+	{0x4367, 0x87},
+	{0x4368, 0x91},
+	{0x4369, 0x8F},
+	{0x436A, 0x8F},
+	{0x436B, 0x8E},
+	{0x436C, 0x93},
+	{0x436D, 0x8D},
+	{0x436E, 0x8E},
+	{0x436F, 0x8D},
+	{0x4370, 0x89},
+	{0x4371, 0x87},
+	{0x4372, 0x87},
+	{0x4373, 0x86},
+	{0x4374, 0x83},
+	{0x4375, 0x83},
+	{0x4376, 0x83},
+	{0x4377, 0x82},
+	{0x4378, 0x81},
+	{0x4379, 0x81},
+	{0x437A, 0x81},
+	{0x437B, 0x81},
+	{0x437C, 0x80},
+	{0x437D, 0x80},
+	{0x437E, 0x80},
+	{0x437F, 0x80},
+	{0x4380, 0x81},
+	{0x4381, 0x82},
+	{0x4382, 0x82},
+	{0x4383, 0x81},
+	{0x4384, 0x84},
+	{0x4385, 0x84},
+	{0x4386, 0x84},
+	{0x4387, 0x82},
+	{0x4388, 0x8A},
+	{0x4389, 0x89},
+	{0x438A, 0x89},
+	{0x438B, 0x87},
+	{0x438C, 0x93},
+	{0x438D, 0x90},
+	{0x438E, 0x90},
+	{0x438F, 0x8E},
+	{0x4390, 0x94},
+	{0x4391, 0x8F},
+	{0x4392, 0x8F},
+	{0x4393, 0x8F},
+	{0x4394, 0x8B},
+	{0x4395, 0x89},
+	{0x4396, 0x89},
+	{0x4397, 0x89},
+	{0x4398, 0x85},
+	{0x4399, 0x85},
+	{0x439A, 0x85},
+	{0x439B, 0x84},
+	{0x439C, 0x82},
+	{0x439D, 0x83},
+	{0x439E, 0x83},
+	{0x439F, 0x82},
+	{0x43A0, 0x81},
+	{0x43A1, 0x82},
+	{0x43A2, 0x82},
+	{0x43A3, 0x81},
+	{0x43A4, 0x82},
+	{0x43A5, 0x83},
+	{0x43A6, 0x84},
+	{0x43A7, 0x82},
+	{0x43A8, 0x86},
+	{0x43A9, 0x86},
+	{0x43AA, 0x86},
+	{0x43AB, 0x84},
+	{0x43AC, 0x8C},
+	{0x43AD, 0x8C},
+	{0x43AE, 0x8C},
+	{0x43AF, 0x89},
+	{0x43B0, 0x96},
+	{0x43B1, 0x94},
+	{0x43B2, 0x95},
+	{0x43B3, 0x91},
+	{0x43B4, 0x99},
+	{0x43B5, 0x93},
+	{0x43B6, 0x94},
+	{0x43B7, 0x94},
+	{0x43B8, 0x90},
+	{0x43B9, 0x8D},
+	{0x43BA, 0x8E},
+	{0x43BB, 0x8D},
+	{0x43BC, 0x89},
+	{0x43BD, 0x89},
+	{0x43BE, 0x8A},
+	{0x43BF, 0x89},
+	{0x43C0, 0x86},
+	{0x43C1, 0x87},
+	{0x43C2, 0x87},
+	{0x43C3, 0x86},
+	{0x43C4, 0x83},
+	{0x43C5, 0x86},
+	{0x43C6, 0x85},
+	{0x43C7, 0x84},
+	{0x43C8, 0x84},
+	{0x43C9, 0x86},
+	{0x43CA, 0x86},
+	{0x43CB, 0x85},
+	{0x43CC, 0x89},
+	{0x43CD, 0x8A},
+	{0x43CE, 0x8A},
+	{0x43CF, 0x87},
+	{0x43D0, 0x90},
+	{0x43D1, 0x91},
+	{0x43D2, 0x91},
+	{0x43D3, 0x8D},
+	{0x43D4, 0x9A},
+	{0x43D5, 0x98},
+	{0x43D6, 0x99},
+	{0x43D7, 0x94},
+	{0x34C0, 0xE4},
+	{0x34C1, 0x00},
+	{0x34C2, 0xF0},
+	{0x34C3, 0x00},
+	{0x34C4, 0xEC},
+	{0x34C5, 0x00},
+	{0x34C6, 0xE9},
+	{0x34C7, 0x00},
+	{0x34C8, 0x8D},
+	{0x34C9, 0x19},
+	{0x34CA, 0xC3},
+	{0x34CB, 0x19},
+	{0x34CC, 0x54},
+	{0x34CD, 0x19},
+	{0x34CE, 0xFF},
+	{0x34CF, 0x18},
+	{0x3053, 0x00},
+	{0x3630, 0x40},
+	{0x3631, 0x00},
+	{0x3632, 0x40},
+	{0x3633, 0x00},
+	{0x3634, 0x40},
+	{0x3635, 0x00},
+	{0x3636, 0x40},
+	{0x3637, 0x00},
+	{0x3638, 0x40},
+	{0x3639, 0x00},
+	{0x363A, 0x40},
+	{0x363B, 0x00},
+	{0x363C, 0x40},
+	{0x363D, 0x00},
+	{0x363E, 0x40},
+	{0x363F, 0x00},
+	{0x3838, 0xA0},
+	{0x3839, 0x00},
+	{0x383A, 0xC6},
+	{0x383B, 0x00},
+	{0x383C, 0x95},
+	{0x383D, 0x00},
+	{0x383E, 0xB2},
+	{0x383F, 0x00},
+	{0x3840, 0xD2},
+	{0x3841, 0xFF},
+	{0x3842, 0xD6},
+	{0x3843, 0xFF},
+	{0x3844, 0xEF},
+	{0x3845, 0xFF},
+	{0x3846, 0xEC},
+	{0x3847, 0xFF},
+	{0x3848, 0x06},
+	{0x3849, 0x00},
+	{0x384A, 0x04},
+	{0x384B, 0x00},
+	{0x384C, 0x06},
+	{0x384D, 0x00},
+	{0x384E, 0xF6},
+	{0x384F, 0xFF},
+};
+
+static const struct imx390_reg_list lsc_369_list = {
+	.num_of_regs = ARRAY_SIZE(imx390_lsc_pattern_369),
+	.regs = imx390_lsc_pattern_369,
 };
 
 static const s64 link_freq_menu_items[] = {
@@ -1338,8 +1839,11 @@ static int imx390_set_lsc_pattern(struct imx390 *self, int val)
 	if (val == LSC_PATTERN_UNITY)
 		ret = imx390_write_reg_list(self, &lsc_unity_list);
 
-	if (val == LSC_PATTERN_TABLE)
-		ret = imx390_write_reg_list(self, &lsc_vendor_def_list);
+	if (val == LSC_PATTERN_TABLE_958)
+		ret = imx390_write_reg_list(self, &lsc_958_list);
+
+	if (val == LSC_PATTERN_TABLE_369)
+		ret = imx390_write_reg_list(self, &lsc_369_list);
 
 	if (!ret)
 		dev_dbg(&client->dev,
-- 
2.17.1

